### ISSUE TITLE
[clang] qualifier discard warning is not emitted when -Wpointer-sign is disabled

### ISSUE THREAD

**Issue by vasama** [llvm/llvm-project#225368]
This code produces a `-Wpointer-sign` warning. When that warning is disabled, no warning is issued, despite the fact that a qualifier is also discarded. GCC issues the expected warning in this case.

```c
unsigned char* f(char const* p)
{
    return p;
}
```

See it here on Compiler Explorer: https://godbolt.org/z/exxEa7doa

**Comment by anandkvinu** [comment]
Can I track this down and try to fix it?

**Comment by shafik** [comment]
So this is b/c gcc is diagnosing the issue under a different flag, if you add `-Wno-discarded-qualifiers` to gcc then it also does not diagnose: https://godbolt.org/z/4rWvj58nr

Maybe I am misunderstanding the issue.

**Comment by vasama** [comment]
Sorry, perhaps I was not clear enough. In the conversion from `char const*` to `unsigned char*`, there are two issues worth diagnosing: the signedness conversion and the discarding of qualifiers. Even when I disable one diagnostic, I still want to see the other.

Looks like GCC has the same issue, just in inverse. It prefers to warn about the discarding of qualifiers, and when that warning is disabled, forgets about the signedness.

**Comment by shafik** [comment]
If I change the example to:

```c
char* f(char const* p)
{
    return p;
}

```

clang does warn:

> error: returning 'const char *' from a function with result type 'char *' discards qualifiers [-Werror,-Wincompatible-pointer-types-discards-qualifiers]

So you just need the right flags.

**Comment by vasama** [comment]
I am aware that Clang is capable of warning about the discarding of qualifiers. The issue is that in the conversion I showed, there are *two* problems worth warning about, and when one is suppressed the other gets lost. Showing how warnings do work in some *other* piece of code is not particularly helpful.
